File: python/promptspec/exceptions/prompt_exceptions.py
"""
Custom exceptions for the PromptSpec library.
"""


class PromptValidationException(Exception):
    """Exception thrown when prompt validation fails."""
    pass


class MissingPlaceholderException(PromptValidationException):
    """Exception thrown when a required placeholder is missing."""
    
    def __init__(self, placeholder_name: str):
        self.placeholder_name = placeholder_name
        super().__init__(f"Required placeholder '{placeholder_name}' is missing.")


class PlaceholderTypeException(PromptValidationException):
    """Exception thrown when a placeholder value has incorrect type."""
    
    def __init__(self, placeholder_name: str, expected_type: str, actual_type: str):
        self.placeholder_name = placeholder_name
        self.expected_type = expected_type
        self.actual_type = actual_type
        super().__init__(
            f"Placeholder '{placeholder_name}' expected type '{expected_type}' "
            f"but received '{actual_type}'."
        )


class PromptNotFoundException(Exception):
    """Exception thrown when a prompt template is not found."""
    
    def __init__(self, prompt_name: str):
        self.prompt_name = prompt_name
        super().__init__(f"Prompt template '{prompt_name}' not found.")

File: src/Exceptions/PromptExceptions.cs

using System;

namespace PromptSpec.Exceptions
{

    /// <summary>
    /// Exception thrown when prompt validation fails.
    /// </summary>
    public class PromptValidationException : Exception
    {
        public PromptValidationException(string message) : base(message) { }

        public PromptValidationException(string message, Exception innerException) : base(message, innerException) { }
    }
    /// <summary>
    /// Exception thrown when a required placeholder is missing.
    /// </summary>
    public class MissingPlaceholderException : PromptValidationException
    {
        public string PlaceholderName { get; }

        public MissingPlaceholderException(string placeholderName)
            : base($"Required placeholder '{placeholderName}' is missing.")
        {
            PlaceholderName = placeholderName;
        }
    }


    /// <summary>
    /// Exception thrown when a placeholder value has an incorrect type.
    /// </summary>
    public class PlaceholderTypeException : PromptValidationException
    {
        public string PlaceholderName { get; }
        public string ExpectedType { get; }
        public string ActualType { get; }

        public PlaceholderTypeException(string placeholderName, string expectedType, string actualType)
            : base($"Placeholder '{placeholderName}' expected type '{expectedType}' but received '{actualType}'.")
        {
            PlaceholderName = placeholderName;
            ExpectedType = expectedType;
            ActualType = actualType;
        }
    }

    /// <summary>
    /// Exception thrown when a prompt template is not found.
    /// </summary>
    public class PromptNotFoundException : Exception
    {
        public string PromptName { get; }

        public PromptNotFoundException(string promptName)
            : base($"Prompt template '{promptName}' not found.")
        {
            PromptName = promptName;
        }
    }

}
